lib/IRGen/GenEnum.cpp

```diff
@@ -659,7 +659,15 @@ namespace {
     void consume(IRGenFunction &IGF, Explosion &src,
                  Atomicity atomicity,
                  SILType T) const override {
-      if (tryEmitConsumeUsingDeinit(IGF, src, T)) {
+      if (ElementsAreABIAccessible &&
```

Comment:
The use of `ElementsAreABIAccessible` here is surprising to me.  I would have assumed that that is set iff all of the elements are ABI-accessible, and I can't see why that figures into the decision about whether to emit a call to `deinit` at the level of `T`.  (We shouldn't be entering this function at all if `T` is itself ABI-inaccessible — we need to have made the decision at a higher level.)

Once we've decided that we don't have a `deinit` at the level of `T`, then yeah, we should check `ElementsAreABIAccessible` before proceeding into the logic that breaks apart the enum.

You have a similar check for all the other aggregate uses, and I think the same logic applies: you don't want to gate the initial `tryEmitConsumeUsingDeinit`, just the splitting-up logic.

Reply:
>we can recurse to the sub elements.
s/we can/we cannot/